core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -414,6 +414,18 @@ public PlanWithProperties visitTopN(TopNNode node, PreferredProperties preferred
                     return rebaseAndDeriveProperties(node, child);
                 case PARTIAL:
                     child = planChild(node, PreferredProperties.any());
+                    // If source is pre-sorted, partial topN can be replaced with partial limit N.
+                    // We record this in isInputPresorted flag in TopNNode instead of replacing with LimitNode
+                    // to avoid pushdown of such replaced LimitNode through the source which was producing ordered input.
+                    List<LocalProperty<Symbol>> desiredProperties = node.getOrderingScheme().getOrderBy().stream()
```

Comment:
There are few places in this class where we convert ordering scheme to sorting properties. Extract a utility method maybe?

Reply:
Added OrderingScheme#toLocalProperties